File: Backend/internal/repositories/health_repository.go
package repositories

import (
	"context"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"pawpal-backend/internal/models"
)

// HealthRepository handles health records and journals database operations
type HealthRepository struct {
	db *pgxpool.Pool
}

// NewHealthRepository creates a new HealthRepository
func NewHealthRepository(db *pgxpool.Pool) *HealthRepository {
	return &HealthRepository{db: db}
}

// Health Record Methods

// CreateHealthRecord creates a new health record
func (r *HealthRepository) CreateHealthRecord(ctx context.Context, record *models.HealthRecord) error {
	query := `
		INSERT INTO health_records (
			id, pet_id, owner_id, is_vaccinated, vaccination_date, vaccination_details,
			medical_conditions, allergies, medications, vet_name, vet_clinic, vet_phone,
			vet_address, emergency_contact_name, emergency_contact_phone, insurance_provider,
			insurance_policy_number, additional_notes, created_at, updated_at
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20)
		RETURNING id, created_at, updated_at`

	now := time.Now()
	record.ID = uuid.New()
	record.CreatedAt = now
	record.UpdatedAt = now

	return r.db.QueryRow(ctx, query,
		record.ID,
		record.PetID,
		record.OwnerID,
		record.IsVaccinated,
		record.VaccinationDate,
		record.VaccinationDetails,
		record.MedicalConditions,
		record.Allergies,
		record.Medications,
		record.VetName,
		record.VetClinic,
		record.VetPhone,
		record.VetAddress,
		record.EmergencyContactName,
		record.EmergencyContactPhone,
		record.InsuranceProvider,
		record.InsurancePolicyNumber,
		record.AdditionalNotes,
		record.CreatedAt,
		record.UpdatedAt,
	).Scan(&record.ID, &record.CreatedAt, &record.UpdatedAt)
}

// GetHealthRecordByPetID gets the health record for a pet
func (r *HealthRepository) GetHealthRecordByPetID(ctx context.Context, petID uuid.UUID, ownerID uuid.UUID) (*models.HealthRecord, error) {
	query := `
		SELECT id, pet_id, owner_id, is_vaccinated, vaccination_date, vaccination_details,
			medical_conditions, allergies, medications, vet_name, vet_clinic, vet_phone,
			vet_address, emergency_contact_name, emergency_contact_phone, insurance_provider,
			insurance_policy_number, additional_notes, created_at, updated_at
		FROM health_records WHERE pet_id = $1 AND owner_id = $2 LIMIT 1`

	record := &models.HealthRecord{}
	err := r.db.QueryRow(ctx, query, petID, ownerID).Scan(
		&record.ID,
		&record.PetID,
		&record.OwnerID,
		&record.IsVaccinated,
		&record.VaccinationDate,
		&record.VaccinationDetails,
		&record.MedicalConditions,
		&record.Allergies,
		&record.Medications,
		&record.VetName,
		&record.VetClinic,
		&record.VetPhone,
		&record.VetAddress,
		&record.EmergencyContactName,
		&record.EmergencyContactPhone,
		&record.InsuranceProvider,
		&record.InsurancePolicyNumber,
		&record.AdditionalNotes,
		&record.CreatedAt,
		&record.UpdatedAt,
	)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}
	return record, nil
}

// GetHealthRecordByID gets a health record by ID
func (r *HealthRepository) GetHealthRecordByID(ctx context.Context, id uuid.UUID) (*models.HealthRecord, error) {
	query := `
		SELECT id, pet_id, owner_id, is_vaccinated, vaccination_date, vaccination_details,
			medical_conditions, allergies, medications, vet_name, vet_clinic, vet_phone,
			vet_address, emergency_contact_name, emergency_contact_phone, insurance_provider,
			insurance_policy_number, additional_notes, created_at, updated_at
		FROM health_records WHERE id = $1`

	record := &models.HealthRecord{}
	err := r.db.QueryRow(ctx, query, id).Scan(
		&record.ID,
		&record.PetID,
		&record.OwnerID,
		&record.IsVaccinated,
		&record.VaccinationDate,
		&record.VaccinationDetails,
		&record.MedicalConditions,
		&record.Allergies,
		&record.Medications,
		&record.VetName,
		&record.VetClinic,
		&record.VetPhone,
		&record.VetAddress,
		&record.EmergencyContactName,
		&record.EmergencyContactPhone,
		&record.InsuranceProvider,
		&record.InsurancePolicyNumber,
		&record.AdditionalNotes,
		&record.CreatedAt,
		&record.UpdatedAt,
	)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}
	return record, nil
}

// UpdateHealthRecord updates a health record
func (r *HealthRepository) UpdateHealthRecord(ctx context.Context, record *models.HealthRecord) error {
	query := `
		UPDATE health_records SET
			is_vaccinated = $2, vaccination_date = $3, vaccination_details = $4,
			medical_conditions = $5, allergies = $6, medications = $7, vet_name = $8,
			vet_clinic = $9, vet_phone = $10, vet_address = $11, emergency_contact_name = $12,
			emergency_contact_phone = $13, insurance_provider = $14, insurance_policy_number = $15,
			additional_notes = $16, updated_at = $17
		WHERE id = $1`

	_, err := r.db.Exec(ctx, query,
		record.ID,
		record.IsVaccinated,
		record.VaccinationDate,
		record.VaccinationDetails,
		record.MedicalConditions,
		record.Allergies,
		record.Medications,
		record.VetName,
		record.VetClinic,
		record.VetPhone,
		record.VetAddress,
		record.EmergencyContactName,
		record.EmergencyContactPhone,
		record.InsuranceProvider,
		record.InsurancePolicyNumber,
		record.AdditionalNotes,
		time.Now(),
	)
	return err
}

// DeleteHealthRecord deletes a health record
func (r *HealthRepository) DeleteHealthRecord(ctx context.Context, id uuid.UUID) error {
	query := `DELETE FROM health_records WHERE id = $1`
	_, err := r.db.Exec(ctx, query, id)
	return err
}

// Health Journal Methods

// CreateHealthJournal creates a new health journal entry
func (r *HealthRepository) CreateHealthJournal(ctx context.Context, journal *models.HealthJournal) error {
	query := `
		INSERT INTO health_journals (
			id, pet_id, owner_id, date, weight, weight_unit, activity_level, energy_level,
			mood, appetite, symptoms, medications_taken, vet_visit, vet_visit_reason,
			vet_notes, general_notes, created_at, updated_at
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18)
		RETURNING id, created_at, updated_at`

	now := time.Now()
	journal.ID = uuid.New()
	journal.CreatedAt = now
	journal.UpdatedAt = now

	return r.db.QueryRow(ctx, query,
		journal.ID,
		journal.PetID,
		journal.OwnerID,
		journal.Date,
		journal.Weight,
		journal.WeightUnit,
		journal.ActivityLevel,
		journal.EnergyLevel,
		journal.Mood,
		journal.Appetite,
		journal.Symptoms,
		journal.MedicationsTaken,
		journal.VetVisit,
		journal.VetVisitReason,
		journal.VetNotes,
		journal.GeneralNotes,
		journal.CreatedAt,
		journal.UpdatedAt,
	).Scan(&journal.ID, &journal.CreatedAt, &journal.UpdatedAt)
}

// GetHealthJournalsByPetID gets all health journal entries for a pet
func (r *HealthRepository) GetHealthJournalsByPetID(ctx context.Context, petID uuid.UUID, ownerID uuid.UUID) ([]models.HealthJournal, error) {
	query := `
		SELECT id, pet_id, owner_id, date, weight, weight_unit, activity_level, energy_level,
			mood, appetite, symptoms, medications_taken, vet_visit, vet_visit_reason,
			vet_notes, general_notes, created_at, updated_at
		FROM health_journals WHERE pet_id = $1 AND owner_id = $2 ORDER BY date DESC`

	rows, err := r.db.Query(ctx, query, petID, ownerID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var journals []models.HealthJournal
	for rows.Next() {
		journal := models.HealthJournal{}
		err := rows.Scan(
			&journal.ID,
			&journal.PetID,
			&journal.OwnerID,
			&journal.Date,
			&journal.Weight,
			&journal.WeightUnit,
			&journal.ActivityLevel,
			&journal.EnergyLevel,
			&journal.Mood,
			&journal.Appetite,
			&journal.Symptoms,
			&journal.MedicationsTaken,
			&journal.VetVisit,
			&journal.VetVisitReason,
			&journal.VetNotes,
			&journal.GeneralNotes,
			&journal.CreatedAt,
			&journal.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}
		journals = append(journals, journal)
	}
	return journals, nil
}

// GetHealthJournalByID gets a health journal entry by ID
func (r *HealthRepository) GetHealthJournalByID(ctx context.Context, id uuid.UUID) (*models.HealthJournal, error) {
	query := `
		SELECT id, pet_id, owner_id, date, weight, weight_unit, activity_level, energy_level,
			mood, appetite, symptoms, medications_taken, vet_visit, vet_visit_reason,
			vet_notes, general_notes, created_at, updated_at
		FROM health_journals WHERE id = $1`

	journal := &models.HealthJournal{}
	err := r.db.QueryRow(ctx, query, id).Scan(
		&journal.ID,
		&journal.PetID,
		&journal.OwnerID,
		&journal.Date,
		&journal.Weight,
		&journal.WeightUnit,
		&journal.ActivityLevel,
		&journal.EnergyLevel,
		&journal.Mood,
		&journal.Appetite,
		&journal.Symptoms,
		&journal.MedicationsTaken,
		&journal.VetVisit,
		&journal.VetVisitReason,
		&journal.VetNotes,
		&journal.GeneralNotes,
		&journal.CreatedAt,
		&journal.UpdatedAt,
	)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}
	return journal, nil
}

// UpdateHealthJournal updates a health journal entry
func (r *HealthRepository) UpdateHealthJournal(ctx context.Context, journal *models.HealthJournal) error {
	query := `
		UPDATE health_journals SET
			weight = $2, weight_unit = $3, activity_level = $4, energy_level = $5,
			mood = $6, appetite = $7, symptoms = $8, medications_taken = $9,
			vet_visit = $10, vet_visit_reason = $11, vet_notes = $12, general_notes = $13, updated_at = $14
		WHERE id = $1`

	_, err := r.db.Exec(ctx, query,
		journal.ID,
		journal.Weight,
		journal.WeightUnit,
		journal.ActivityLevel,
		journal.EnergyLevel,
		journal.Mood,
		journal.Appetite,
		journal.Symptoms,
		journal.MedicationsTaken,
		journal.VetVisit,
		journal.VetVisitReason,
		journal.VetNotes,
		journal.GeneralNotes,
		time.Now(),
	)
	return err
}

// DeleteHealthJournal deletes a health journal entry
func (r *HealthRepository) DeleteHealthJournal(ctx context.Context, id uuid.UUID) error {
	query := `DELETE FROM health_journals WHERE id = $1`
	_, err := r.db.Exec(ctx, query, id)
	return err
}

File: App/lib/features/pets/data/repositories/health_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import '../models/health_record_model.dart';
import '../models/health_journal_model.dart';

class HealthRepository {
  final FirebaseFirestore _firestore;
  final FirebaseAuth _auth;

  HealthRepository({
    FirebaseFirestore? firestore,
    FirebaseAuth? auth,
  })  : _firestore = firestore ?? FirebaseFirestore.instance,
        _auth = auth ?? FirebaseAuth.instance;

  /// Get current user ID
  String? get _currentUserId => _auth.currentUser?.uid;

  /// Reference to health records collection
  CollectionReference get _healthRecordsCollection =>
      _firestore.collection('health_records');

  /// Reference to health journals collection
  CollectionReference get _healthJournalsCollection =>
      _firestore.collection('health_journals');

  /// Create health record for a pet
  Future<String?> createHealthRecord({
    required String petId,
    required bool isVaccinated,
    String? vaccinationDate,
    String? vaccinationDetails,
    List<String>? medicalConditions,
    List<String>? allergies,
    List<String>? medications,
    String? vetName,
    String? vetClinic,
    String? vetPhone,
    String? vetAddress,
    String? emergencyContactName,
    String? emergencyContactPhone,
    String? insuranceProvider,
    String? insurancePolicyNumber,
    String? additionalNotes,
  }) async {
    try {
      final userId = _currentUserId;
      if (userId == null) throw Exception('User not authenticated');

      // Generate health record ID
      final doc = _healthRecordsCollection.doc();
      final recordId = doc.id;

      // Create health record model
      final healthRecord = HealthRecordModel(
        id: recordId,
        petId: petId,
        ownerId: userId,
        isVaccinated: isVaccinated,
        vaccinationDate: vaccinationDate,
        vaccinationDetails: vaccinationDetails,
        medicalConditions: medicalConditions,
        allergies: allergies,
        medications: medications,
        vetName: vetName,
        vetClinic: vetClinic,
        vetPhone: vetPhone,
        vetAddress: vetAddress,
        emergencyContactName: emergencyContactName,
        emergencyContactPhone: emergencyContactPhone,
        insuranceProvider: insuranceProvider,
        insurancePolicyNumber: insurancePolicyNumber,
        additionalNotes: additionalNotes,
        createdAt: DateTime.now(),
        updatedAt: DateTime.now(),
      );

      // Save to Firestore
      await doc.set(healthRecord.toJson());

      return recordId;
    } catch (e) {
      print('Error creating health record: $e');
      return null;
    }
  }

  /// Get health record for a pet
  Future<HealthRecordModel?> getHealthRecord(String petId) async {
    try {
      final userId = _currentUserId;
      if (userId == null) return null;

      final snapshot = await _healthRecordsCollection
          .where('petId', isEqualTo: petId)
          .where('ownerId', isEqualTo: userId)
          .limit(1)
          .get();

      if (snapshot.docs.isEmpty) return null;

      return HealthRecordModel.fromJson(
          snapshot.docs.first.data() as Map<String, dynamic>);
    } catch (e) {
      print('Error getting health record: $e');
      return null;
    }
  }

  /// Update health record
  Future<bool> updateHealthRecord({
    required String recordId,
    bool? isVaccinated,
    String? vaccinationDate,
    String? vaccinationDetails,
    List<String>? medicalConditions,
    List<String>? allergies,
    List<String>? medications,
    String? vetName,
    String? vetClinic,
    String? vetPhone,
    String? vetAddress,
    String? emergencyContactName,
    String? emergencyContactPhone,
    String? insuranceProvider,
    String? insurancePolicyNumber,
    String? additionalNotes,
  }) async {
    try {
      final userId = _currentUserId;
      if (userId == null) throw Exception('User not authenticated');

      final updateData = <String, dynamic>{
        'updatedAt': FieldValue.serverTimestamp(),
      };

      if (isVaccinated != null) updateData['isVaccinated'] = isVaccinated;
      if (vaccinationDate != null) updateData['vaccinationDate'] = vaccinationDate;
      if (vaccinationDetails != null) updateData['vaccinationDetails'] = vaccinationDetails;
      if (medicalConditions != null) updateData['medicalConditions'] = medicalConditions;
      if (allergies != null) updateData['allergies'] = allergies;
      if (medications != null) updateData['medications'] = medications;
      if (vetName != null) updateData['vetName'] = vetName;
      if (vetClinic != null) updateData['vetClinic'] = vetClinic;
      if (vetPhone != null) updateData['vetPhone'] = vetPhone;
      if (vetAddress != null) updateData['vetAddress'] = vetAddress;
      if (emergencyContactName != null) updateData['emergencyContactName'] = emergencyContactName;
      if (emergencyContactPhone != null) updateData['emergencyContactPhone'] = emergencyContactPhone;
      if (insuranceProvider != null) updateData['insuranceProvider'] = insuranceProvider;
      if (insurancePolicyNumber != null) updateData['insurancePolicyNumber'] = insurancePolicyNumber;
      if (additionalNotes != null) updateData['additionalNotes'] = additionalNotes;

      await _healthRecordsCollection.doc(recordId).update(updateData);
      return true;
    } catch (e) {
      print('Error updating health record: $e');
      return false;
    }
  }

  /// Create health journal entry
  Future<String?> createHealthJournalEntry({
    required String petId,
    required DateTime date,
    double? weight,
    String? weightUnit,
    String? activityLevel,
    String? energyLevel,
    String? mood,
    String? appetite,
    List<String>? symptoms,
    List<String>? medicationsTaken,
    bool? vetVisit,
    String? vetVisitReason,
    String? vetNotes,
    String? generalNotes,
  }) async {
    try {
      final userId = _currentUserId;
      if (userId == null) throw Exception('User not authenticated');

      // Generate journal entry ID
      final doc = _healthJournalsCollection.doc();
      final entryId = doc.id;

      // Create health journal model
      final journalEntry = HealthJournalModel(
        id: entryId,
        petId: petId,
        ownerId: userId,
        date: date,
        weight: weight,
        weightUnit: weightUnit,
        activityLevel: activityLevel,
        energyLevel: energyLevel,
        mood: mood,
        appetite: appetite,
        symptoms: symptoms,
        medicationsTaken: medicationsTaken,
        vetVisit: vetVisit,
        vetVisitReason: vetVisitReason,
        vetNotes: vetNotes,
        generalNotes: generalNotes,
        createdAt: DateTime.now(),
        updatedAt: DateTime.now(),
      );

      // Save to Firestore
      await doc.set(journalEntry.toJson());

      return entryId;
    } catch (e) {
      print('Error creating health journal entry: $e');
      return null;
    }
  }

  /// Get health journal entries for a pet
  Stream<List<HealthJournalModel>> getHealthJournalEntries(String petId) {
    final userId = _currentUserId;
    if (userId == null) {
      return Stream.value([]);
    }

    return _healthJournalsCollection
        .where('petId', isEqualTo: petId)
        .where('ownerId', isEqualTo: userId)
        .orderBy('date', descending: true)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs
          .map((doc) => HealthJournalModel.fromJson(doc.data() as Map<String, dynamic>))
          .toList();
    });
  }

  /// Get health journal entry by ID
  Future<HealthJournalModel?> getHealthJournalEntry(String entryId) async {
    try {
      final doc = await _healthJournalsCollection.doc(entryId).get();
      if (!doc.exists) return null;

      return HealthJournalModel.fromJson(doc.data() as Map<String, dynamic>);
    } catch (e) {
      print('Error getting health journal entry: $e');
      return null;
    }
  }

  /// Update health journal entry
  Future<bool> updateHealthJournalEntry({
    required String entryId,
    double? weight,
    String? weightUnit,
    String? activityLevel,
    String? energyLevel,
    String? mood,
    String? appetite,
    List<String>? symptoms,
    List<String>? medicationsTaken,
    bool? vetVisit,
    String? vetVisitReason,
    String? vetNotes,
    String? generalNotes,
  }) async {
    try {
      final userId = _currentUserId;
      if (userId == null) throw Exception('User not authenticated');

      final updateData = <String, dynamic>{
        'updatedAt': FieldValue.serverTimestamp(),
      };

      if (weight != null) updateData['weight'] = weight;
      if (weightUnit != null) updateData['weightUnit'] = weightUnit;
      if (activityLevel != null) updateData['activityLevel'] = activityLevel;
      if (energyLevel != null) updateData['energyLevel'] = energyLevel;
      if (mood != null) updateData['mood'] = mood;
      if (appetite != null) updateData['appetite'] = appetite;
      if (symptoms != null) updateData['symptoms'] = symptoms;
      if (medicationsTaken != null) updateData['medicationsTaken'] = medicationsTaken;
      if (vetVisit != null) updateData['vetVisit'] = vetVisit;
      if (vetVisitReason != null) updateData['vetVisitReason'] = vetVisitReason;
      if (vetNotes != null) updateData['vetNotes'] = vetNotes;
      if (generalNotes != null) updateData['generalNotes'] = generalNotes;

      await _healthJournalsCollection.doc(entryId).update(updateData);
      return true;
    } catch (e) {
      print('Error updating health journal entry: $e');
      return false;
    }
  }

  /// Delete health journal entry
  Future<bool> deleteHealthJournalEntry(String entryId) async {
    try {
      final userId = _currentUserId;
      if (userId == null) throw Exception('User not authenticated');

      await _healthJournalsCollection.doc(entryId).delete();
      return true;
    } catch (e) {
      print('Error deleting health journal entry: $e');
      return false;
    }
  }

  /// Get health journal entries for a specific date range
  Future<List<HealthJournalModel>> getHealthJournalEntriesInRange(
    String petId,
    DateTime startDate,
    DateTime endDate,
  ) async {
    try {
      final userId = _currentUserId;
      if (userId == null) return [];

      final snapshot = await _healthJournalsCollection
          .where('petId', isEqualTo: petId)
          .where('ownerId', isEqualTo: userId)
          .where('date', isGreaterThanOrEqualTo: startDate)
          .where('date', isLessThanOrEqualTo: endDate)
          .orderBy('date', descending: true)
          .get();

      return snapshot.docs
          .map((doc) => HealthJournalModel.fromJson(doc.data() as Map<String, dynamic>))
          .toList();
    } catch (e) {
      print('Error getting health journal entries in range: $e');
      return [];
    }
  }
}